dlls/dxgi/factory.c

```diff
@@ -658,17 +659,59 @@ static HRESULT STDMETHODCALLTYPE dxgi_factory_CreateSwapChainForHwnd(IDXGIFactor
 {
     struct qemu_dxgi_factory_CreateSwapChainForHwnd call;
     struct qemu_dxgi_factory *factory = impl_from_IDXGIFactory5(iface);
+    HRESULT hr;
+    struct qemu_dxgi_device *device_impl;
+    struct qemu_dxgi_swapchain *obj;
+    uint64_t *host_buffers;
 
     call.super.id = QEMU_SYSCALL_ID(CALL_DXGI_FACTORY_CREATESWAPCHAINFORHWND);
     call.iface = (ULONG_PTR)factory;
-    call.device = (ULONG_PTR)device;
     call.window = (ULONG_PTR)window;
+    call.swapchain = (ULONG_PTR)swapchain;
     call.swapchain_desc = (ULONG_PTR)swapchain_desc;
     call.fullscreen_desc = (ULONG_PTR)fullscreen_desc;
-    call.output = (ULONG_PTR)output;
-    call.swapchain = (ULONG_PTR)swapchain;
+
+    if (device)
+    {
+        device_impl = unsafe_impl_from_IDXGIDevice(device);
+        if (!device)
+        {
+            WINE_WARN("This is not the device we are looking for.\n");
+            return E_FAIL;
+        }
+        call.device = (ULONG_PTR)device_impl;
+    }
+    else
+    {
+        call.device = 0;
+        device_impl = NULL;
+    }
+
+    if (output)
+    {
+        call.output = (ULONG_PTR)unsafe_impl_from_IDXGIOutput((IUnknown *)output);
```

Comment:
I just wonder why I made unsafe_impl_from_IDXGIOutput accept an IUnknown * instead of an IDXGIOutput * ...